marathon/push_listener.go

```diff
@@ -0,0 +1,147 @@
+package marathon
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type Event struct {
+	EventType  string    `json:"eventType"`
+	Timestamp  time.Time `json:"timestamp"`
+	SlaveID    string    `json:"slaveId"`
+	TaskID     string    `json:"taskId"`
+	TaskStatus string    `json:"taskStatus"`
+	AppID      string    `json:"appId"`
+	Host       string    `json:"host"`
+	Ports      []int     `json:"ports"`
+	Version    string    `json:"version"`
+}
+
+type Listener struct {
+	events       chan Event
+	host         string
+	internalPort string // Internal/external ports are relative
+	externalPort string // to the container this process runs in.
+}
+
+func NewListener(host string, internalPort, externalPort string) *Listener {
+	listener := &Listener{
+		events:       make(chan Event),
+		host:         host,
+		internalPort: internalPort,
+		externalPort: externalPort,
+	}
+	http.HandleFunc("/push-listener", listener.handler)
+	go http.ListenAndServe(":"+internalPort, nil)
+
+	return listener
+}
+
+func (l *Listener) handler(res http.ResponseWriter, req *http.Request) {
+	decoder := json.NewDecoder(req.Body)
+
+	var event Event
+	if err := decoder.Decode(&event); err != nil {
+		log.Fatal(err)
+	}
+
+	if event.EventType == "status_update_event" { // We only care about container change events
+		l.events <- event
+	}
+
+	res.Write([]byte("Thanks.")) // Marathon ignores replies.  Just being polite.
+}
+
+func (l *Listener) Events() <-chan Event {
+	return l.events
+}
+
+func (l *Listener) unsubscribe(marathonHost, callback string) error {
+	marathonURL := url.URL{Scheme: "http", Host: marathonHost, Path: "/v2/eventSubscriptions"}
+	q := marathonURL.Query()
+	q.Set("callbackUrl", callback)
+	marathonURL.RawQuery = q.Encode()
+
+	req, err := http.NewRequest("DELETE", marathonURL.String(), nil)
+	res, err := http.DefaultClient.Do(req)
+	defer res.Body.Close()
```

Comment:
I think that this potentially causes a panic in the case that there is an error (as res.Body is not guaranteed to be a valid object in that case).